cardano-crypto-class/src/Cardano/Crypto/DSIGN/Ed25519.hs

```diff
@@ -0,0 +1,105 @@
+{-# LANGUAGE DeriveGeneric #-}
+{-# LANGUAGE DerivingStrategies #-}
+{-# LANGUAGE DerivingVia #-}
+{-# LANGUAGE FlexibleInstances #-}
+{-# LANGUAGE GeneralizedNewtypeDeriving #-}
+{-# LANGUAGE ScopedTypeVariables #-}
+{-# LANGUAGE TypeFamilies #-}
+
+-- | Ed25519 digital signatures.
+module Cardano.Crypto.DSIGN.Ed25519
+  ( Ed25519DSIGN
+  , SigDSIGN (..)
+  , SignKeyDSIGN (..)
+  , VerKeyDSIGN (..)
+  )
+where
+
+import Cardano.Binary
+  ( Decoder
+  , Encoding
+  , FromCBOR (..)
+  , ToCBOR (..)
+  , serialize
+  )
+import Cardano.Crypto.DSIGN.Class
+import Cardano.Prelude (NoUnexpectedThunks, UseIsNormalForm(..))
+import Crypto.Error (CryptoFailable (..))
+import Crypto.PubKey.Ed25519
+import Data.ByteArray (ByteArrayAccess, convert)
+import Data.ByteString (ByteString)
+import Data.ByteString.Lazy (toStrict)
+import GHC.Generics (Generic)
+
+data Ed25519DSIGN
+
+instance DSIGNAlgorithm Ed25519DSIGN where
+
+    type Signable Ed25519DSIGN = ToCBOR
+
+    newtype VerKeyDSIGN Ed25519DSIGN = VerKeyEd25519DSIGN PublicKey
+        deriving (Show, Eq, Generic, ByteArrayAccess)
+        deriving NoUnexpectedThunks via UseIsNormalForm PublicKey
+
+    newtype SignKeyDSIGN Ed25519DSIGN = SignKeyEd25519DSIGN SecretKey
+        deriving (Show, Eq, Generic, ByteArrayAccess)
+        deriving NoUnexpectedThunks via UseIsNormalForm SecretKey
+
+    newtype SigDSIGN Ed25519DSIGN = SigEd25519DSIGN Signature
+        deriving (Show, Eq, Generic, ByteArrayAccess)
+        deriving NoUnexpectedThunks via UseIsNormalForm Signature
+
+    encodeVerKeyDSIGN = toCBOR
+    encodeSignKeyDSIGN = toCBOR
+    encodeSigDSIGN = toCBOR
+
+    decodeVerKeyDSIGN = fromCBOR
+    decodeSignKeyDSIGN = fromCBOR
+    decodeSigDSIGN = fromCBOR
+
+    genKeyDSIGN = SignKeyEd25519DSIGN <$> generateSecretKey
+
+    deriveVerKeyDSIGN (SignKeyEd25519DSIGN sk) = VerKeyEd25519DSIGN $ toPublic sk
+
+    signDSIGN () a (SignKeyEd25519DSIGN sk) = do
+        let vk = toPublic sk
+            bs = toStrict $ serialize a
+        return $ SigEd25519DSIGN $ sign sk vk bs
+
+    verifyDSIGN () (VerKeyEd25519DSIGN vk) a (SigEd25519DSIGN sig) =
+        if verify vk (toStrict $ serialize a) sig
```

Comment:
And then instead of `serialise` here is where we should recover the annotated bytes for the serialised rep.

Similarly for `serialise` in `signDSIGN`.

Reply:
As per previous comment: the annotated instances in Shelley provide that `serialise` does _not_ result in re-serialisation.